## Fantasy Baseball Auction Value ##

__This notebook provides the content for today's lecture, as well as an introduction to Assignment 4. The details for Assignment 4 are provided at the end of the notebook.__

#### Due: Wednesday, March 20 by 3pm. ####

Hello everyone, not Rhonda here for once (it's your CAs, Ashna and Brandon). As you've probably heard by now, we're going to be doing a fantasy baseball league in this class. The object of this assignment is to come up with some way for evaluating players. How you might ask? However you want we might reply. You've heard about WAR and wOBA and replacement value and all sorts of ways to evaluate players, and now we're going to turn it into a competition with (fake) money on it (and a prize at the end). 

#### How does fantasy baseball work?

Fantasy baseball assigns point values to different events for real like players, such as runs, homeruns, innings pitched, etc. You can see how your leagues scoring and rosters are set up in the section below under "You mentioned scoring?". At the end of the competition (near the end of the semester), whatever team has the most points in each league is the winner! Ashna and I will be playing as well to give you all some competition.

__The winning teams get to choose the type of cake we order on the last day of class.__

#### How do we pick our team?

In fantasy baseball, each player can only belong to one team (just like in real life). You will be selecting your team with a partner through an **auction draft**. In an auction draft, you'll start with \\$250. You'll take turns nominating a player, and then everyone will live bid on that player, just like a real life auction. The least a player can go for is \\$1, while the most is \\$250-(# roster spots - 1). You'll have to use your money wisely to build the best team possible, however you can still use waivers and trades to keep your roster competitive.

#### What are waivers?

Waivers are where any non-rostered player is at. You can pick up players from waivers, but you'll have to drop a player from your roster to do so.

#### You mentioned scoring?

| Batting Stat | Points | 
|--------------|--------|
| AB           | -1.0   |
| H            | 5.6    |
| 2B           | 2.9    |
| 3B           | 5.7    |
| HR           | 9.4    |
| BB           | 3.0    |
| HBP          | 3.0    |
| SB           | 1.9    |
| CS           | -2.8   |  
  
| Pitching Stat | Points |
|---------------|--------|
| IP            | 5      |
| K             | 2.0    |
| BB            | -3.0   |
| HBP           | -3.0   |
| HR            | -13    |
| SV            | 5.0    |
| HLD           | 4.0    |  
  
| Position | Roster Spots |
|----------|--------------|
| C        | 1            |
| 1B       | 1            |
| 2B       | 1            |
| 3B       | 1            |
| SS       | 1            |
| OF       | 3            |
| UTIL     | 1            |
| SP       | 5            |
| RP       | 2            |
| Bench    | 3            |  
  
A few notes about these settings. ESPN counts the "IP" stat as per out, so in the scoring settings it is 5/3=1.66, however per every other statkeeping reference, it's 5 points per IP. A 2B/3B/HR is added on to the value of a hit, so a double is worth 5.6+2.9=8.5 points. These weights are based on wOBA/FIP, adjusted to keep pitchers and batter on a reatively level playing field. OF is any outfield position. UTIL is any batter, doesn't matter the position (kind of like a DH).

# Assignment 4 - Fantasy Baseball #

### Join a league ###

The first step in this assignment is to join a fantasy league. There are two leagues you can join that are setup specifically for this class. Both have identcal settings, so pick one.   
http://fantasy.espn.com/baseball/league/join?leagueId=90287999&inviteId=471eac27-e559-4737-8f3e-6c13f054a6a1
http://fantasy.espn.com/baseball/league/join?leagueId=5441683&inviteId=cf19e2aa-38d9-4886-8aa6-9465aeb58369  
You'll need an ESPN account (free). In order to add your group-mate so that you can both control your team, go to "My Team" at the top of the page, and then under your team name, there's a "Add 2nd Owner" link. If you're having trouble with this, let one of Ashna or I know and we'll add you ourselves. 

### Design an approach to valuing players ###

Glad you asked. Now you get to put your real baseball skills to work. Your assignment is to come up with a way to value players. You can use FanGraphs projections and base them off of that. You can look at historical data and create your own prediction model. You can use some fancy neural network or machine learning algorithm if you want to go the extra mile. It's a pretty open ended question, just at the end you need to come up with some sort of a list that says "Player A is work \\$X" for every fantasy relevant player in the league. 

#### Not entirely sure where to get started? ####

That's fair, it's a pretty open ended task, especially if you've never done fantasy sports or an auction draft before. Here are some links to help you learn the ropes:  
  
[FanGraphs Fantasy Auction Calculator.](https://www.fangraphs.com/auctiontool.aspx) Now you can't just export this and turn it in, but it can be a helpful tool for knowing a general range for values.  
  
[A basic autcion draft overview.](https://757fantasyleague.wordpress.com/auction-draft-the-basics/) This can help you get more familiar with how an auction draft works and some of the basic strategies. You can't get every player that you want, so you need to budget you money.  
  
Personally, I would recommend coming up with a way to evaluate players based on the categories that we use for scoring. This should net you some expected points. Again, this can come from whatever math or form you like. Then worry about turning those point values into auction dollar values, and there are a number of ways to do that (you can do some Googling). 

#### How about some sweet hints?

No worries, in order to help even out the playing field a bit, here are some hints that can help you in the draft.  
1. Try to get other people to spend their money on players that you don't want but you think that other people will. You can use ESPN draft rankings you help find discrepencies between your rankings and the ones that other people will see to help with that.  
2. There's more than one strategy to building a team. You can blow all your money on some top tier talent and fill in the rest of your team with cheap players, or build a more balanced roster. Both can work.  
3. There's position scarcity in baseball (some positions are deeper than others), meaning that the worst 2B on a roster might put up 1.9 points a game while the worst 1B on a roster might put up 2.5. Establishing a replacement level *wink wink* for each position can help you guage which players are more valuable just becuase of the position they play. 
4. Fantasy value does not always equal real life value. We don't care about defense and players have to play to score points. There might be some great guy sitting on an MLB teams bench, but he's not going to give you fantasy points if he's only playing every third day. Here, counting stats matter.  
5. The league is only going to last a little over a month (the last day of points scoring will be 11:59pm on April 28th), meaning that prospects or players that won't be called up until midway through the season (Vlad Jr., Eloy) aren't realy worth drafting. You're welcome to ask if you don't know who those players are.

### Submit your code and writeup ###

The product that you produce for this assignment is a notebook that describes your approach, as well as the code that implements the approach. Your notebook needs to include the following sections and information:

1. Introduction - A summary of your general approach
2. Data and statistics - Describe the data and statistics that you used, and how they were used.
3. Results of your approach - This section needs to include the top 5 players that your approach identifies. You also need to select one player from your top 5 and provide an analysis of his rating and monetary value.
4. Limitations - What difficulties do you see with implementing your strategy in the draft?

The draft will occur in class on March 20. Your assignment needs to be submitted before class on that day. 

To make grading easier, please name the notebook *Assignment4_LastName1_LastName2*, where *LastName1* and *LastName2* are the last names of the team members. Yes, you will be working in teams of two for this assignment.

## Samuel Reed & Joe Wilson Assignment 4 ##

In [33]:
import pandas as pd
C_data = pd.read_csv('C_data.csv')
B1_data = pd.read_csv('1B_data.csv')
B2_data = pd.read_csv('2B_data.csv')
B3_data = pd.read_csv('3B_data.csv')
SS_data = pd.read_csv('SS_data.csv')
RF_data = pd.read_csv('RF_data.csv')
CF_data = pd.read_csv('CF_data.csv')
LF_data = pd.read_csv('LF_data.csv')
SP_data = pd.read_csv('SP_data.csv')
RP_data = pd.read_csv('RP_data.csv')

In [34]:
avg_C_stats = pd.DataFrame({"C": (C_data[['AB', 'H','2B', '3B', 'HR', 'BB', 'HBP', 'SB', 'CS' ]].mean())})
avg_1B_stats = pd.DataFrame({"1B": (B1_data[['AB', 'H','2B', '3B', 'HR', 'BB', 'HBP', 'SB', 'CS' ]].mean())})
avg_2B_stats = pd.DataFrame({"2B": (B2_data[['AB', 'H','2B', '3B', 'HR', 'BB', 'HBP', 'SB', 'CS' ]].mean())})
avg_3B_stats = pd.DataFrame({"3B": B3_data[['AB', 'H','2B', '3B', 'HR', 'BB', 'HBP', 'SB', 'CS' ]].mean()})
avg_SS_stats = pd.DataFrame({"SS": SS_data[['AB', 'H','2B', '3B', 'HR', 'BB', 'HBP', 'SB', 'CS' ]].mean()})
avg_RF_stats = pd.DataFrame({"RF": (RF_data[['AB', 'H','2B', '3B', 'HR', 'BB', 'HBP', 'SB', 'CS' ]].mean())})
avg_CF_stats = pd.DataFrame({"CF": (CF_data[['AB', 'H','2B', '3B', 'HR', 'BB', 'HBP', 'SB', 'CS' ]].mean())})
avg_LF_stats = pd.DataFrame({"LF": (LF_data[['AB', 'H','2B', '3B', 'HR', 'BB', 'HBP', 'SB', 'CS' ]].mean())})
avg_SP_stats = pd.DataFrame({"SP": SP_data[['IP', 'SO','BB', 'HBP', 'HR', 'SV', 'HLD']].mean()})
avg_RP_stats = pd.DataFrame({"RP": RP_data[['IP', 'SO','BB', 'HBP', 'HR', 'SV', 'HLD']].mean()})

avg_stats = pd.concat([avg_C_stats, avg_1B_stats, avg_2B_stats, avg_3B_stats, avg_SS_stats], axis=1)
avg_stats2 = pd.concat([avg_RF_stats, avg_CF_stats, avg_LF_stats], axis=1)
avg_stats_batters = pd.concat([avg_stats, avg_stats2], axis=1)
avg_stats_batters.loc['score'] = (avg_stats_batters.loc['AB'] * (-1)) + (avg_stats_batters.loc['H'] * 5.6) + (avg_stats_batters.loc['2B'] * 2.9) + (avg_stats_batters.loc['3B'] * 5.7) + (avg_stats_batters.loc['HR'] * 9.4) + (avg_stats_batters.loc['BB'] * 3.0) + (avg_stats_batters.loc['HBP'] * 3.0) + (avg_stats_batters.loc['SB'] * 1.9) + (avg_stats_batters.loc['CS'] * (-2.8))
avg_stats_batters.loc['dollar/point'] = 13.15 / avg_stats_batters.loc['score']
print("Average Stats for Batters of Specific Positions")
avg_stats_batters

Average Stats for Batters of Specific Positions


C          1B          2B          3B          SS  \
AB            336.918919  429.041667  439.931818  427.595745  458.268293   
H              81.864865  107.937500  115.613636  110.702128  119.731707   
2B             16.513514   21.833333   22.659091   24.872340   24.707317   
3B              1.081081    1.604167    2.727273    1.957447    2.902439   
HR             11.216216   18.062500   12.772727   17.170213   14.219512   
BB             30.864865   47.895833   38.227273   43.276596   35.926829   
HBP             6.000000    4.770833    4.204545    5.276596    4.439024   
SB              1.270270    2.895833    8.704545    4.127660   10.756098   
CS              1.027027    1.229167    3.272727    1.808511    4.292683   
score         391.140541  577.716667  543.495455  585.461702  563.602439   
dollar/point    0.033620    0.022762    0.024195    0.022461    0.023332   

                      RF          CF          LF  
AB            407.259259  440.090909  425.117647  
H             106.925926  112.590909  110.921569  
2B             22.611111   21.613636   21.490196  
3B              2.425926    3.795455    2.686275  
HR             15.185185   13.704545   16.137255  
BB             45.666667   44.659091   44.176471  
HBP             5.222222    5.795455    4.627451  
SB              6.574074   13.477273    6.921569  
CS              2.666667    4.772727    2.803922  
score         571.357407  567.161364  577.078431  
dollar/point    0.023015    0.023186    0.022787

In [35]:
avg_stats_pitchers = pd.concat([avg_SP_stats, avg_RP_stats], axis=1)
avg_stats_pitchers.loc['score'] = (avg_stats_pitchers.loc['IP'] * 5) + (avg_stats_pitchers.loc['SO'] * 2) + (avg_stats_pitchers.loc['BB'] * (-3)) + (avg_stats_pitchers.loc['HBP'] * (-3)) + (avg_stats_pitchers.loc['HR'] * (-13)) + (avg_stats_pitchers.loc['SV'] * 5) + (avg_stats_pitchers.loc['HLD'] * 4)
avg_stats_pitchers.loc['dollar/point'] = 13.15 / avg_stats_pitchers.loc['score']
print("Average Stats for Starting Pitchers and Relief Pitchers")
avg_stats_pitchers

Average Stats for Starting Pitchers and Relief Pitchers


SP          RP
IP            183.055738   69.392135
SO            178.950820   73.134831
BB             54.967213   24.516854
HBP             7.262295    2.910112
HR             22.311475    6.820225
SV              0.000000    8.067416
HLD             0.000000   13.483146
score         796.442623  416.556180
dollar/point    0.016511    0.031568

The score row at the bottom of these data frames represents the average ESPN fantasy score for a player at that position last year. These scores would represent what a $13.15 player at that position would look like statistically.  

Under that is the dollar/point row. This represents how many dollars an increase of one fantasy point at that position is worth, given that an average player is worth $13.15. 

In [36]:
C_total = pd.read_csv('Pos_total/C_total.csv')
B1_total = pd.read_csv('Pos_total/1B_total.csv')
B2_total = pd.read_csv('Pos_total/2B_total.csv')
B3_total = pd.read_csv('Pos_total/3B_total.csv')
SS_total = pd.read_csv('Pos_total/SS_total.csv')
RF_total = pd.read_csv('Pos_total/RF_total.csv')
CF_total = pd.read_csv('Pos_total/CF_total.csv')
LF_total = pd.read_csv('Pos_total/LF_total.csv')
SP_total = pd.read_csv('Pos_total/SP_total.csv')
RP_total = pd.read_csv('Pos_total/RP_total.csv')

In [37]:
def calc_dollars_batter(player_name, pos):
    if pos == 'C':
        pos_data = C_total
    elif pos == '1B':
        pos_data = B1_total
    elif pos == '2B':
        pos_data = B2_total
    elif pos == '3B':
        pos_data = B3_total
    elif pos == 'SS':
        pos_data = SS_total
    elif pos == 'RF':
        pos_data = RF_total
    elif pos == 'CF':
        pos_data = CF_total
    elif pos == 'LF':
        pos_data = LF_total
    
    player_data = pos_data.loc[pos_data['Name'] == player_name]
    score = (player_data['AB'] * (-1)) + (player_data['H'] * 5.6) + (player_data['2B'] * 2.9) + (player_data['3B'] * 5.7) + (player_data['HR'] * 9.4) + (player_data['BB'] * 3.0) + (player_data['HBP'] * 3.0) + (player_data['SB'] * 1.9) + (player_data['CS'] * (-2.8))
    point_diff = float(score - avg_stats_batters[pos]['score'])
    if point_diff < 0:
        return 13.15 - (abs(point_diff) * avg_stats_batters[pos]['dollar/point'])
    elif point_diff > 0:
        return 13.15 + (point_diff * avg_stats_batters[pos]['dollar/point'])
    else:
        return 13.15

In [38]:
calc_dollars_batter("Mike Trout", "CF")

28.91713196901635

In [39]:
def calc_dollars_pitcher(player_name, pos):
    if pos == 'SP':
        pos_data = SP_total
    elif pos == 'RP':
        pos_data = RP_total
    
    player_data = pos_data.loc[pos_data['Name'] == player_name]
    score = (player_data['IP'] * 5) + (player_data['SO'] * 2) + (player_data['BB'] * (-3)) + (player_data['HBP'] * (-3)) + (player_data['HR'] * (-13)) + (player_data['SV'] * 5) + (player_data['HLD'] * 4)
    point_diff = float(score - avg_stats_pitchers[pos]['score'])
    if point_diff < 0:
        return 13.15 - (abs(point_diff) * avg_stats_pitchers[pos]['dollar/point'])
    elif point_diff > 0:
        return 13.15 + (point_diff * avg_stats_pitchers[pos]['dollar/point'])
    else:
        return 13.15

In [40]:
calc_dollars_pitcher("Bryan Shaw", "RP")

7.166020203110043

## Introduction ##
<p>
    Looking at a players performance through the lens of statistics provides a different perspective than a subjective approach normally would. In order to guide our drafting strategty we decided to tabulate lists of what the average player in every position looks like statistically. This way we can look at the statistics of a specific player compared to an average player and guage how much they are worth in monetary terms based off of their statistics. We chose an approach like this because it will allow us to make monetarily intelligent decisions about what players we pick. If we bid for an above average player, we will know what that player is worth compared to other players statisically and will be able to bid accordingly without over-valuing or under-valuing a player. With a budget of 250 Dollars and 19 positions to fill our average player will need to cost roughly 13 dollars. 
</p>

## Data and Statistics ##
<p>
    The first step in shaping our drafting strategy was to collect all of the relevant data and process what the average player in each position looks like statistically. In our data collection we only focused on the data points that were repesented in the overall scoring calculation for a player and we used data from the 2018 regular season. Those stats are AB, H, 2B, 3B, HR, BB, HBP, SB, and CS for batters and IP, SO, BB, HBP, HR, SV, and HLD for pitchers. We also only collected stats on batters with more than 250 plate appearances and pitchers with more than 160 innings pitched so that we our data would be representive of an average player with a significant amount of play time during the 2018 season. Below are our results of what an average player in every position will look like statistically. The score row at the bottom of these data frames represents the average ESPN fantasy score for a player at that position last year. These scores would represent what a 13.15 Dollar player at that position would look like statistically. Under that is the dollar/point row. This represents how many dollars an increase of one fantasy point at that position is worth, given that an average player is worth 13.15.
</p>

In [41]:
avg_stats_batters

C          1B          2B          3B          SS  \
AB            336.918919  429.041667  439.931818  427.595745  458.268293   
H              81.864865  107.937500  115.613636  110.702128  119.731707   
2B             16.513514   21.833333   22.659091   24.872340   24.707317   
3B              1.081081    1.604167    2.727273    1.957447    2.902439   
HR             11.216216   18.062500   12.772727   17.170213   14.219512   
BB             30.864865   47.895833   38.227273   43.276596   35.926829   
HBP             6.000000    4.770833    4.204545    5.276596    4.439024   
SB              1.270270    2.895833    8.704545    4.127660   10.756098   
CS              1.027027    1.229167    3.272727    1.808511    4.292683   
score         391.140541  577.716667  543.495455  585.461702  563.602439   
dollar/point    0.033620    0.022762    0.024195    0.022461    0.023332   

                      RF          CF          LF  
AB            407.259259  440.090909  425.117647  
H             106.925926  112.590909  110.921569  
2B             22.611111   21.613636   21.490196  
3B              2.425926    3.795455    2.686275  
HR             15.185185   13.704545   16.137255  
BB             45.666667   44.659091   44.176471  
HBP             5.222222    5.795455    4.627451  
SB              6.574074   13.477273    6.921569  
CS              2.666667    4.772727    2.803922  
score         571.357407  567.161364  577.078431  
dollar/point    0.023015    0.023186    0.022787

In [42]:
avg_stats_pitchers

SP          RP
IP            183.055738   69.392135
SO            178.950820   73.134831
BB             54.967213   24.516854
HBP             7.262295    2.910112
HR             22.311475    6.820225
SV              0.000000    8.067416
HLD             0.000000   13.483146
score         796.442623  416.556180
dollar/point    0.016511    0.031568

## Results ##
<p>
The results of this approach are slightly atypical because they do not point to a specific set of players for the ideal roster. Instead this approach leverages the fact that in an auction style draft a team is unlikely to get all of the players the planned out in advance, and our draft will let us evaluate players on the fly so we can see how much they are worth. To do this we have created functions that let us enter a player's name and position and the function will return a dollar amount of how much that player is worth based on how much above or below the average their statistics are. 
<p>

In [43]:
print(calc_dollars_batter("Josh Donaldson", "3B"))
print(calc_dollars_pitcher("Chris Sale", "SP"))

6.360928454356807
16.13116830990264


This allows us to identify the top 5 pitchers and top 5 batters. The code and results of that investigation are below:

In [44]:
C_data['dollars'] = C_data['Name'].apply(lambda name: calc_dollars_batter(name, 'C'))
B1_data['dollars'] = B1_data['Name'].apply(lambda name: calc_dollars_batter(name, '1B'))
B2_data['dollars'] = B2_data['Name'].apply(lambda name: calc_dollars_batter(name, '2B'))
B3_data['dollars'] = B3_data['Name'].apply(lambda name: calc_dollars_batter(name, '3B'))
SS_data['dollars'] = SS_data['Name'].apply(lambda name: calc_dollars_batter(name, 'SS'))
LF_data['dollars'] = LF_data['Name'].apply(lambda name: calc_dollars_batter(name, 'LF'))
CF_data['dollars'] = CF_data['Name'].apply(lambda name: calc_dollars_batter(name, 'CF'))
RF_data['dollars'] = RF_data['Name'].apply(lambda name: calc_dollars_batter(name, 'RF'))

SP_data['dollars'] = SP_data['Name'].apply(lambda name: calc_dollars_pitcher(name, 'SP'))
RP_data['dollars'] = RP_data['Name'].apply(lambda name: calc_dollars_pitcher(name, 'RP'))

In [45]:
batters = pd.concat([C_data, B1_data, B2_data, B3_data, SS_data, LF_data, CF_data, RF_data])
batters.nlargest(6, 'dollars')

Name     Team    G   AB   PA    H   1B  2B  3B  HR    ...      \
0       Mookie Betts  Red Sox  136  520  614  180   96  47   5  32    ...       
0         Mike Trout   Angels  140  471  608  147   80  24   4  39    ...       
1   Christian Yelich  Brewers  147  574  651  187  110  34   7  36    ...       
0   Christian Yelich  Brewers  147  574  651  187  110  34   7  36    ...       
17      Jose Ramirez  Indians  157  578  698  156   75  38   4  39    ...       
1      Manny Machado    - - -  162  632  709  188  113  35   3  37    ...       

     SO  HBP  SF  SH  GDP  SB  CS    AVG  playerid    dollars  
0    91    8   5   0    5  30   6  0.346     13611  29.017774  
0   124   10   4   0    5  24   2  0.312     10155  28.917132  
1   135    7   2   0   14  22   4  0.326     11477  27.749627  
0   135    7   2   0   14  22   4  0.326     11477  27.474524  
17   80    8   6   0    2  34   6  0.270     13510  26.616173  
1   104    2   5   0   26  14   2  0.297     11493  26.229890  

[6 rows x 24 columns]

In [46]:
pitchers = pd.concat([SP_data, RP_data])
pitchers.nlargest(5, 'dollars')

Name       Team   W  L   ERA   G  GS  CG  ShO  SV    ...      ER  \
3      Edwin Diaz   Mariners   0  4  1.96  73   0   0    0  57    ...      16   
0    Jacob deGrom       Mets  10  9  1.70  32  32   1    0   0    ...      41   
0   Blake Treinen  Athletics   9  2  0.78  68   0   0    0  38    ...       7   
5    Max Scherzer  Nationals  18  7  2.53  33  33   2    1   0    ...      62   
14     Josh Hader    Brewers   6  1  2.43  55   0   0    0  12    ...      22   

    HR  BB  IBB  HBP  WP  BK   SO  playerid    dollars  
3    5  17    0    6   3   1  124     14710  24.134020  
0   10  46    3    5   2   0  269     10954  22.124632  
0    2  21    3    1   6   0  100     12572  22.050507  
5   23  51    4   12   4   1  300      3137  20.027745  
14   9  30    0    1   0   1  143     14212  19.746016  

[5 rows x 26 columns]

We included the top 6 players in the batters list because Christian Yelich appeared in our left fielder and right fielder data. In general, the most valuable position players are worth more than the most valuable pitchers.  

Mookie Betts is the most valuable player out of all the data from last season. Our function puts his value at about \$29, more than double the average player. This makes perfect sense. Aside from being one of the coolest guys in Major League Baseball, he also won the AL MVP award last year. He hit .346/.438/.640 last season. His WAR clocked in at 10.9 on Baseball Reference, first place amongst position players. He is a classic 5-tool player: he has a great arm from a difficult right field at Fenway Park, his defensive abilities in the outfield are nothing short of stellar (especially considering he's a natural second baseman), he can hit for power as his 32 homeruns show, he can hit for contact, and let's not forget he's incredibly fast. Mookie Betts being at the top of our list of most valuable players is a sign that our evaluation function is working properly.

Despite our on the fly drafting approach we have created a few rosters that are composed of statisically valuable players and how much we plan on paying for those players. Below are some of our sample rosters that would be suitable for a team based on our approach

    C: 
        Jonathan Lucroy ~11.5
        Max Stassi ~8.7
        Buster Posey - 16.65
        Willson Contreras - 19.8
    1B:
        Mitch Moreland ~12
        Yuli Gurriel ~14
        Joey Votto - ~19.52
        Anthony Rizzo ~21.05
    2B:
        DJ LeMahieu ~15.7 (but I think we could get a deal)
        Dee Gordon ~11.8
        Daniel Murphy - 10.71 (good deal)
        Robinson Cano - 11.69 (good deal)
    SS:
        Dansby Swanson ~12.4
        Nick Ahmed ~13
        Troy Tulowitzki (Doesn’t have a value cause didn’t play last season but I have a feeling he might surprise people. We’d be able to get him for probably nothing. Good bench candidate.)
        Trea Turner - 21.31
    3B:
        Justin Turner ~15.2
        Matt Chapman ~19.8 (expensive but might be able to get a deal)
        Nolan Arenado - 24.5
        Josh Donaldson - 6.3
    LF:
        Corey Dickerson ~15.6
        Marwin Gonzalez ~13.50 (slightly underrated I’d say)
    CF:
        Aaron Hicks ~18.6
        Kevin Pillar ~13.2
    RF:
        Mookie Betts ~29 (probably way too expensive but he’s my favorite player)
        Max Kepler ~15
        Gregory Polanco ~17.2
    SP:
        Trevor Bauer ~16
        Gerrit Cole ~18
        Blake Snell ~15.5
        J.A. Happ ~12.2
        Jake Arrieta ~11
        Mike Fiers ~9.7
        Matthew Boyd ~10
        Kyle Freeland ~15
        Kyle Hendricks ~14
    RP:
        Heath Hembree ~10
        Buck Farmer ~8
        Brandon Kintzler ~10
        Ryan Brasier ~6
        Bobby Poyner ~3



## Limitations ##
<p>
    This approach provides a solid foundation for quickly evaluating the monetary worth of a player on the fly, but that in and of itself has some limitations. By basing all of our decisions off the average player and how a player we are evaluating compares to the average player, we may miss opportunities to pay high amounts of money for star players who will go on to out perform what we estimate them to monetarily worth. This approach is also limited in the fact we must think and make decisions very quickly, and if we spend an above average amount of money on a player we will need to keep track of how much we have spent and buy below average players to make sure we can afford to fill all of our positions.
</p>

In [48]:
# helper function to help us keep track of how much we have left to spend
# add the price of a player to player_prices every time we purchase a player
player_prices = []
def get_budget(arr):
    sum = 0
    for item in arr:
        sum += item
    return sum

spent = get_budget(player_prices)
remaining = 250 - spent
avg_cost = remaining/19-len(player_prices)
print("We have spent", spent, "and we have", remaining, "remaning leaving us with an average player price of", avg_cost)

We have spent 0 and we have 250 remaning leaving us with an average player price of 13.157894736842104
